### ISSUE TITLE
redis subscriber discovery

### ISSUE THREAD

**Issue by nflaig** [rahulrkr08/loopback4-redis-pubsub#2]
Hi @rahulrkr08  👋 

I like your component but how exactly does the RedisPubSubProvider discovery work?

As far as I can just adding the `redisSubscriber` decorator won't be enough to bind it to the application.

You either would have to do something like `this.add(createBindingFromClass(TestRedisPubSubProvider));` in application.ts or add it to the services folder and follow the convention described in [Service Booter](https://loopback.io/doc/en/lb4/Booting-an-Application.html#service-booter).

I am guessing you are using the component yourself in one of your applications would be interesting to know which pattern you are suggesting. Or maybe it could even be considered to add a [custom booter](https://loopback.io/doc/en/lb4/Booting-an-Application.html#custom-booters) for this in the component itself.

**Comment by rahulrkr08** [comment]
Hi @nflaig,

Thanks for your query, first of all, apologies for the late reply.

`RedisPubSubComponent` uses [extensionPoint](https://loopback.io/doc/en/lb4/Extension-point-and-extensions.html) which attach services, life cycle observers etc to the context. The `redisSubscriber` decorator returns a callback with bind function, which actually attaches the provider class (custom provider class). So, the time when the loopback is booting up it will bind the provider class with the current context.

When the `start()` ([Life cycle](https://loopback.io/doc/en/lb4/Life-cycle.html)) function from `RedisPubSubObserver` is getting called, it retrives all the provider classes tagged to `RedisPubSubBindings.REDIS_PUBSUB_NAMESPACE` along with its channel name, and assign to the `on` subcribe function of redis client, which will call the handler function defined in provider class (`value()`)

As the `redisSubscriber` is attaching the provider class to the application context, we not required to have `this.add(createBindingFromClass(TestRedisPubSubProvider))`. I made the design in such a way that, the user should not add any custom code apart from attaching the component and define the provider class along with the decorator. In case, if we change the design `this.add(createBindingFromClass(TestRedisPubSubProvider))` this function needs to call for multiple times to subscribe to multiple channels if they have.

I Hope, you have cleared your doubts.

**Comment by nflaig** [comment]
Hi @rahulrkr08 

Thanks for the response.

I did some testing with the component and it works as expected but for the redis subscriber to be automatically bound to the application I had to put the file into the `services` folder and use the extension `.service.ts`. That was the expected behavior as far as I can see if I put the file in any other folder in my application the `TestRedisPubSubProvider` provider will not be bound.
